#!/usr/bin/env perl

use strict;
use warnings;

use Getopt::Long;
use Iterator::FastaDb;
use Iterator::FastqDb;

my $usage=<<'ENDHERE';
NAME:
generateMaskFromRNAmmer.pl

PURPOSE:
Generate a GTF file containing coordinates of rRNA and tRNA
sequences in input files.

INPUT:
--infile_rRNA <string> : Fasta sequence file generated by rnammer.
--infile_tRNA <string> : Fasta sequence file generated by tRNAscan
--infile_gtf <string>  : ...

OUTPUT:
STDOUT (GTF file)

STDOUT

NOTES:

BUGS/LIMITATIONS:
 
AUTHOR/SUPPORT:
Julien Tremblay - julien.tremblay@mail.mcgill.ca

ENDHERE

## OPTIONS
my ($help, $infile_rRNA, $infile_tRNA);
my $verbose = 0;

GetOptions(
    'infile_rRNA=s' => \$infile_rRNA,
    'infile_tRNA=s' => \$infile_tRNA,
    'verbose' 	    => \$verbose,
    'help' 		    => \$help
);
if ($help) { print $usage; exit; }

## MAIN
my $ref_fasta_db = Iterator::FastaDb->new($infile_rRNA) or die("Unable to open Fasta file, $infile_rRNA\n");
while( my $curr = $ref_fasta_db->next_seq() ) {
	my $header = $curr->header();
	$header =~ /rRNA_17_2331448-2333023_DIR+ /molecule=18s_rRNA /score=898.0/;
	$header =~ /0/;
	$counter++;
}

